File: src/audio/AudioBuffer.cpp
#include "AudioBuffer.h"

AudioBuffer::AudioBuffer()
    : sampleRate_(44100), channels_(1) {
}

AudioBuffer::AudioBuffer(int sampleRate, int channels)
    : sampleRate_(sampleRate), channels_(channels) {
}

AudioBuffer::~AudioBuffer() {
}

void AudioBuffer::setData(const std::vector<float>& data) {
    data_ = data;
}

void AudioBuffer::appendData(const std::vector<float>& data) {
    data_.insert(data_.end(), data.begin(), data.end());
}

void AudioBuffer::clear() {
    data_.clear();
}

const std::vector<float>& AudioBuffer::getData() const {
    return data_;
}

std::vector<float>& AudioBuffer::getData() {
    return data_;
}

int AudioBuffer::getSampleRate() const {
    return sampleRate_;
}

int AudioBuffer::getChannels() const {
    return channels_;
}

size_t AudioBuffer::getLength() const {
    return data_.size();
}

float AudioBuffer::getDuration() const {
    if (sampleRate_ == 0 || channels_ == 0) return 0.0f;
    return static_cast<float>(data_.size()) / (sampleRate_ * channels_);
}

void AudioBuffer::setSampleRate(int rate) {
    sampleRate_ = rate;
}

void AudioBuffer::setChannels(int channels) {
    channels_ = channels;
}

// Pitch curve 메타데이터 관리
void AudioBuffer::setPitchCurve(const std::vector<float>& curve) {
    pitchCurve_ = curve;
}

const std::vector<float>& AudioBuffer::getPitchCurve() const {
    return pitchCurve_;
}

bool AudioBuffer::hasPitchCurve() const {
    return !pitchCurve_.empty();
}

void AudioBuffer::clearPitchCurve() {
    pitchCurve_.clear();
}

File: web/js/js/audio/AudioBuffer.js
/**
 * AudioBuffer - 오디오 데이터 관리 클래스 (JavaScript 버전)
 */
export class AudioBuffer {
    constructor(sampleRate = 48000, channels = 1) {
        this.sampleRate = sampleRate;
        this.channels = channels;
        this.data = new Float32Array(0);
    }

    /**
     * 오디오 데이터 설정
     * @param {Float32Array|Array} samples
     */
    setData(samples) {
        if (samples instanceof Float32Array) {
            this.data = samples;
        } else {
            this.data = new Float32Array(samples);
        }
    }

    /**
     * 오디오 데이터 가져오기
     * @returns {Float32Array}
     */
    getData() {
        return this.data;
    }

    /**
     * 샘플 수
     * @returns {number}
     */
    getLength() {
        return this.data.length;
    }

    /**
     * 길이 (초)
     * @returns {number}
     */
    getDuration() {
        return this.data.length / this.sampleRate;
    }

    /**
     * 샘플레이트
     * @returns {number}
     */
    getSampleRate() {
        return this.sampleRate;
    }

    /**
     * 채널 수
     * @returns {number}
     */
    getChannels() {
        return this.channels;
    }

    /**
     * 복사본 생성
     * @returns {AudioBuffer}
     */
    clone() {
        const cloned = new AudioBuffer(this.sampleRate, this.channels);
        cloned.setData(new Float32Array(this.data));
        return cloned;
    }
}
